### ISSUE TITLE
Crashes during sync

### ISSUE THREAD

**Issue by sammy-SC** [couchbase/couchbase-lite-ios#2197]
I’ve released my app with `CouchbaseLite-Swift 2.0.3`. During development and testing I didn’t spot any issues but when released to production to a larger number of users, number of crashes spiked. All of the issues seem to be linked to syncing. These are the 5 crashes that occur in CouchbaseLite-Swift 2.0.3. Each crash has a link to gist with stack trace, name of method where it crashes, number of total crashes and iOS versions distribution.

--------------

### [Crash report 1](https://gist.github.com/sammy-SC/69fe6ae5d15956ab71cb17d690bc638e)

- Crash in `CouchbaseLiteSwift uWS::WebSocketProtocol<false>::consume(char, unsigned int, void) + 44`
- **Total number of crashes**: 18 crashes
- Exception:
```
Crashed: Websocket C4 dispatch
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x0000000000000001
```
- **iOS versions**
  - **11.4.1**: 82%
  - **11.2.5**: 6%
  - **11.4.0**: 6%
  - **11.3.1**: 6%

--------------

### [Crash report 2](https://gist.github.com/sammy-SC/aee25b87372f1f8dea22373836da0a30)

- Crash was introduced with `CouchbaseLite-Swift 2.0.3`, it does not have CouchbaseLiteSwift in stack trace however it seems to related to networking
- **Total number of crashes**: 6 crashes
- Exception:
```
Fatal Exception: std::__1::system_error
mutex lock failed: Invalid argument
```
- **iOS versions**
  - **11.4.1**: 66%
  - **11.4.0**: 17%
  - **11.3.1**: 17%

--------------

### [Crash report 3](https://gist.github.com/sammy-SC/b0e3a0a6f40756542b50035c4f0e0fdb)

- Crash in `CouchbaseLiteSwift litecore::blip::Connection::setWebSocket(litecore::websocket::WebSocket*, fleeceapi::AllocedDict const&) + 4256`
- **Total number of crashes**: 4 crashes
- Exception:
```
Crashed: Websocket C4 dispatch
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x0000000000000048
```
- **iOS versions**
  - **11.4.1**: 50%
  - **11.3.0**: 50%

--------------

### [Crash report 4](https://gist.github.com/sammy-SC/96f84d1d45eca967c916086a6a43f1d0)

- Crashes in `CouchbaseLiteSwift litecore::websocket::WebSocketImpl::onReceive(fleece::slice) + 124`
- **Total number of crashes**: 2 crashes
- Exception:
```
Crashed: Websocket C4 dispatch
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x0000000000000000
```
- **iOS versions**
  - **12.0.0**: 100%

--------------

### [Crash report 5](https://gist.github.com/sammy-SC/def0875e653933a569c9e0c4ea9fdb8e)

- Crashes in `CouchbaseLiteSwift c4socket_gotHTTPResponse + 48`
- **Total number of crashes**: 1 crash
- Exception:
```
Crashed: Websocket C4 dispatch
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x7461747300000001
```
- **iOS versions**
  - **11.4.1**: 100%

----

Version:   CouchbaseLite-Swift 2.0.3
Server:    2.0.0

**Comment by pasin** [comment]
@sammy-SC Thanks a lot for the crash reports. 

Can you also provide exception information (e.g. EXC_CRASH SIGABRT and etc) of each crash as well? Normally the information is at the top at the crash report? 

In 2.1 that we will release soon, we have reimplemented `CBLWebSocket` which should be able to help some of these crashes especially crash report 2.

**Comment by sammy-SC** [comment]
@pasin I've added exception information to each crash.

Is there an expected release date for CouchbaseLite 2.1? 

This issue has been tagged with label `known-issue`, for future reference, where do I find a list of known issues? I can't find it within this repo.

Thank you

**Comment by famanson** [comment]
@pasin hiya, this seems to me like a very serious issue, and `CBLWebSocket` is the core of the sync itself, why is it classed as `icebox`? Why is this not prioritised for a hotfix?

**Comment by pasin** [comment]
We will release 2.1 which has a new CBLWebSocket implementation by the end of this week. This means that some of the crashes here (at least the second report) will be taken care in 2.1.

But this issue and #2196 came very late in the release cycle, we are not able to accommodate them in 2.1 release. We plan to have a patch release (2.1.1 tentatively) to mainly focus on crashes in replication.

**Comment by pasin** [comment]
@sammy-SC I couldn't see the exception detail on each report.

**Comment by sammy-SC** [comment]
@pasin my apologise for not including the exception detail, I thought I did. Exception details should now be visible under each crash.

**Comment by pasin** [comment]
@sammy-SC you can filter Github issue with `known-issue` label and 2.1 milestone to see the list of the known issues.

Sorry, there is no workaround for that at the moment. The try/catch needs to be done at the Lite-Core level. 

FWIW, I did one [fix](b42cf2694db48006e790f28dcdebfec1e401b481) that fixes the`bad_alloc` crash reported in #2196.  You may cherry pick the fix to the `release/2.1.x` branch and build the framework yourself.

Please update the exception info and we will try to fix the issue as soon as we can.

**Comment by famanson** [comment]
hi @pasin, is there any progress on this release?

**Comment by snej** [comment]
@sammy-SC, It's much better to file a separate bug report for each distinct crash. What we have here are up to five bugs, and it's unclear to me from reading the discussion which are fixed in 2.1 and which are still issues (and of those, which may be duplicates.)

**Comment by shobhakartiwari** [comment]
I am using CBLite 2.0 and it getting crashed randomly , kindly help in resolving this issue :

**terminating with uncaught exception of type std::__1: (worry) ystem_error: mutex lock failed: Invalid argument**

**Comment by sammy-SC** [comment]
@snej I assumed those crashes were related. I will close this issue and separate each crash into it's own issue. Sorry for the trouble.

**Comment by djpongh** [comment]
Hopefully we fixed this in 2.1.2 coming out shortly. Please reopen if found again.

**Comment by srinadh-k** [comment]
Hello, i am also getting crash, so any plans when you are release 2.1.2
